c++/src/kj/async-inl.h

```diff
@@ -233,21 +420,52 @@ class PromiseArenaMember {
   // memory, e.g. call `delete this`.
   //
   // We use this instead of a virtual destructor for two reasons:
-  // 1. Coroutine nodes are not indepnedent objects, they have to call destroy() on the coroutine
+  // 1. Coroutine nodes are not independent objects, they have to call destroy() on the coroutine
   //    handle to delete themselves.
   // 2. XThreadEvents sometimes leave it up to a different thread to actually delete the object.
 
+  Maybe<EnvironmentSet&> getEnvironmentSet();
+  // Return this member's arena's environment set, if this is an arena-allocated PromiseArenaMember.
+  //
+  // Notably, PromiseArenaMembers which are allocated by a method other than the `allocPromise()`
+  // function (e.g., statically-allocated, stack-allocated, coroutine-frame-allocated PromiseNodes,
+  // etc.) will not have an arena, and thus have no environment set available.
+
+protected:
+  PromiseArenaMember(Maybe<EnvironmentSet> environmentSet)
+      : environmentSet(kj::mv(environmentSet)) {}
+  // Subclasses which are dynamically allocated by something other than `allocPromise()` should
+  // manually capture the environment and use this constructor. In practice, this means Coroutines
+  // should call this constructor.
+
+  PromiseArenaMember() = default;
+  // All other subclasses should use this constructor. If we're being allocated or appended,
+  // PromiseDisposer::alloc/append() will set our `environmentSet` after construction.
+
 private:
   PromiseArena* arena = nullptr;
   // If non-null, then this PromiseNode is the last node allocated within the given arena, and
   // therefore owns the arena. After this node is destroyed, the arena should be deleted.
   //
   // PromiseNodes are allocated within the arena starting from the end, and `PromiseNode`s
   // allocated this way are required to have `PromiseNode` itself as their leftmost inherited type,
-  // so that the pointers match. Thus, the space in `arena` from its start to the location of the
-  // `PromiseNode` is known to be available for subsequent allocations (which should then take
+  // so that the pointers match. Thus, the space in `arena->space` from its start to the location of
+  // the `PromiseNode` is known to be available for subsequent allocations (which should then take
   // ownership of the arena).
 
+  struct NoEnv {};
+
+  OneOf<NoEnv, Maybe<EnvironmentSet>, Maybe<EnvironmentSet&>> environmentSet = NoEnv();
```

Comment:
This is adding three pointers to the size of every `PromiseArenaMember`. If we could manage to optimize it down to `Maybe<EnvironmentSet&>`, then it would be adding only one pointer, which could be a pretty significant difference. The main thing standing in the way seems to be the "standalone" PromiseArenaMembers that don't end up allocated inside an arena. Looking at this, uh, I suspect this code path already leaks memory before your change (my fault). It looks like they will be allocated on the heap and never freed at present. Presumably we don't today have any PromiseArenaMembers that take this code path, or if we do, we have bigger problems?

Of course, even better would be if this didn't have to be tracked on every single node at all. This might be possible. The purpose of this field is only to set the correct environment during the destructor, right? But I think we can say that it's the parent PromiseNode's responsibility to set the correct environment before destroying children. In fact that's probably largely satisfied already, when a parent PromiseNode exists. The only problem case is when there is no parent, i.e. the PromiseNode is directly held by the application via a Promise. Here, we cannot expect the application to set the correct environment when dropping the Promise. HOWEVER, in this particular case, `arena` should always be non-null! So we can get the environment off of the arena! (Well, expect for the stand-alone promise memory leak issue discussed above... which needs to be solved some other way anyway.)